bluez/audio/telephony-ofono.c

```diff
@@ -1646,38 +1653,49 @@ static gboolean handle_manager_modem_removed(DBusConnection *conn,
 	return TRUE;
 }
 
-static gboolean statefs_batt_update(gpointer data)
+static gboolean statefs_batt_update(gint fd, GIOCondition cond, gpointer data)
 {
-	gboolean ret = FALSE;
-	gchar *buf = NULL;
+	gchar buf[8];
 	gchar *endp = NULL;
-	gsize len = 0;
-	GError *err = NULL;
 	guint64 val;
+	int r;
 
-	if (statefs_batt_path == NULL)
-		goto done;
+	if ((cond & G_IO_ERR) || (cond & G_IO_NVAL) || (cond & G_IO_HUP)) {
+		error("Failed to read battery charge.");
+		goto fail;
+	}
 
-	DBG("Reading battery charge from '%s'.", statefs_batt_path);
-	if (g_file_get_contents(statefs_batt_path, &buf, &len, &err) == FALSE) {
-		DBG("Failed to read battery charge: %s", err->message);
-		goto done;
+	DBG("Reading battery charge from fd %d", statefs_batt_fd);
+	memset(buf, 0, sizeof(buf));
+	r = read(statefs_batt_fd, buf, sizeof(buf) - 1);
+	if (r < 0) {
+		error("Failed to read battery charge: %s (%d)",
+			strerror(errno), errno);
+		goto fail;
+	} else if (r == 0) {
+		error("End of file for fd %d", statefs_batt_fd);
+		goto fail;
 	}
 
+	lseek(statefs_batt_fd, 0, SEEK_SET);
+
+	DBG("Read charge value: '%s'", buf);
 	val = g_ascii_strtoull(buf, &endp, 10);
 	if (endp == NULL || *endp != '\0') {
-		DBG("Cannot process battery charge string '%s'", buf);
-		goto done;
+		error("Cannot process battery charge string '%s'", buf);
+		goto fail;
 	}
 
 	DBG("Battery charge changed to %llu", val);
-	val = 5*val/100;
+	val = MIN(5, 5*(val+10)/100);
```

Comment:
Yes, the battery charge needs to be converted to a value in the range 0..5. 

I modified the previous calculation a bit so that 90..100 would give 5 bars instead of only 100 like before.